cpp/src/Slice/Parser.h

```diff
@@ -796,6 +796,7 @@ namespace Slice
             MetadataList typeMetadata);
         TypePtr type() const;
         MetadataList typeMetadata() const;
+        void setTypeMetadata(MetadataList metadata);
```

Comment:
I have a hard time following this API wrt metadata.

The gist of this PR seems to be that we only "set" metadata on a SyntaxTreeBase element after validating it. However here, the constructor takes a MetadatList named typeMetadata (not validated??). And Sequence has apparently two metadatalist: the plain metadata inherited from Contained, and this type metadata. 

How does all this fit together?

What the rationale for having two different kings of metadata for sequence definitions?

Reply:
I think you're getting caught up on the smaller details that aren't important to the overall picture. But it is:
1) Parsing: The parser parses metadata and stores it as-is (without validating it)
2) Validating: After parsing finished, then we go back and validate the metadata
3) Filtering: If any metadata failed validation, we remove it.

Note that none of this is changing in my PR. In the languages that validate metadata, this is already what they do:

**slice2cpp:** https://github.com/zeroc-ice/ice/blob/a2cbac19b054bd08d8c980d77272da295f9348ce/cpp/src/slice2cpp/Gen.cpp#L1110-L1114 **slice2java:** https://github.com/zeroc-ice/ice/blob/a2cbac19b054bd08d8c980d77272da295f9348ce/cpp/src/slice2java/JavaUtil.cpp#L354-L357 **slice2cs:** https://github.com/zeroc-ice/ice/blob/a2cbac19b054bd08d8c980d77272da295f9348ce/cpp/src/slice2cs/CsUtil.cpp#L2075-L2079 (This one returns void because it calls `setMetadata` itself, instead of letting the caller do it. This means it skips metadata inside sequences, dictionaries, and consts; this is a bug).

This PRs primary goal is to centralize and simplify all these separate validators.

Reply:
But for the sake of completeness!
Here's a full answer to your questions, explaining how/why the parser handles metadata the way it does:

In the Ice compiler, we never use the term 'AST'. But if we did `SyntaxTreeBase` is the base type for everything that lives within the AST.

`Contained` is any type that is 'contained' within a container of some kind. Since all Slice definitions have to appear within a `module`, almost everything inherits from `Contained` except for smaller things like `Comment` and `Metadata`, or special things like `Unit` (the whole compilation job) and `DefinitionContext` (a single file).
Because of this `Contained` is effectively synonymous with "A Slice definition".

Slice supports metadata in lots of places:
- At the top of a Slice file (a `DefinitionContext`)
- In front of a Slice definition (a `Contained`)
- Or in special places like on the inner type of sequences, dictionaries, and constants.

Note that the least common denominator of all these types is `SyntaxTreeBase`.

----

Now, let's say we start up the parser. Bison parses tokens from the inside out.
What I mean is if we take this grammar rule (paraphrased): `definitions : metadata definition`
Before this rule will be called, the parser has _already_ parsed the `metadata` and the `definition` separately.
Only after they've both been parsed, will the 'outer' `definitions` rule be called.

Also, our parser goes from the top of files to the bottom.
So, it's possible that we'll encounter types before they're fully defined yet.

Because of these two facts, when the parser first sees metadata directives, there's information that we don't have access to, simply because the parser hasn't seen the whole file yet, or some outer containers haven't been fully constructed yet.

So. The parser parses the metadata, without performing validation, and stores it, wherever it's supposed to be stored (`DefinitionContext`, `Contained`, `Sequence`, etc).

Then, eventually, the parser finishes. We've now seen the entire file, we have all the information, and everything has been fully constructed. So _now_, with this full picture of everything, we go back and validate the metadata.

During validation, if we find any metadata that is invalid, we retroactively remove it. So that when the code-gen runs, it can assume that all the metadata it sees is correct.

----

And finally, there's the actual metadata validation itself.
In `icerpc`, the AST is a simple `Vec<Node>` that we can iterate through. We don't have this in Ice.
The only way to 'iterate' through Slice definitions is with a `Visitor`.

So, we create a subclass of visitor, that visits every single place where metadata can exist, and calls a function for each place. This function runs the actual validation (so that the validation can be customized by overriding it).

This function needs to take the metadata (since that's what we're validating), and it also needs a reference to what the metadata has been applied to (since `Metadata` doesn't hold this information internally).
And the function needs some way to communicate back to it's caller - whether the metadata was okay.
We could just use a bool. But `optional<string>` lets us provide a nice message too.

Reply:
> Then, eventually, the parser finishes. We've now seen the entire file, we have all the information, and everything has been fully constructed. So now, with this full picture of everything, we go back and validate the metadata.

> During validation, if we find any metadata that is invalid, we retroactively remove it. So that when the code-gen runs, it can assume that all the metadata it sees is correct.

Let's try to be clear here. There is code in the Slice library and code in the Slice compiler.

I assume both perform some level of validation: the code in the Slice library validates `amd` and other non-language specific metadata while slice2cpp validates `cpp:` metadata.

And, this validation is combined in some manner with filtering: we warn about invalid metadata and remove it _before_ the actual code generation. Which means slice2cpp better instantiate and run its metadata visitor (/validator/filter) first.

Questions:
- does it make sense to use a single Visitor instance--created by slice2cpp--to perform the `amd` and `cpp:xxx` metadata validation+filtering? For me, the answer is no. We create many visitors during a compilation; I don't see a reason to optimize the validation+filtering by using a single instance for everything.

- when slice2cpp validates and filters metadata using a visitor, is the visiting API really so inappropriate that we need a different API? And is a single validate function the best API for the job? That's the main question asked by this PR.

And I completely agree we need to figure this out before you update all the compilers.

Reply:
> I assume both perform some level of validation

Currently, no. The "Slice library" performs no validation of metadata. This is something we're definitely going to fix.

> We create many visitors during a compilation; I don't see a reason to optimize the validation+filtering by using a single instance for everything.

My reasoning for combining them isn't about performance (although one iteration is obviously better than 2).
It's about output cleanliness. It's nicer if:
1) We emit all the metadata warnings together, instead of being split between multiple sections.
2) The metadata warnings are all in order (the order they occur in your file).

To be, this is a benefit. And I couldn't come up with a downside to combining them. So I thought Pros > Cons.
Is there a downside to doing this in one pass?

> is the visiting API really so inappropriate that we need a different API?

I think we've moved this discussion elsewhere. Let me know if you still want to carry it in on here.